### PR TITLE
support inplace buffer reuse

### PR DESCRIPTION
**Pull request by jgong5**
Enable `inplace_buffers = True` feature. This addresses the issue https://github.com/pytorch/torchdynamo/issues/823 and the second optimization mentioned in  https://github.com/pytorch/torchdynamo/issues/1401. It is turned off by default now due to https://github.com/pytorch/torchdynamo/issues/1670. Currently, only one CI case gmixer_24_224 training on CUDA would fail due to this issue.

This PR made following fixes:
1. Support more than two `other_names` in `make_inplace` (common.py) to support a chain of inplace reuses.
2. Do not delete buffer if it is the graph output in `make_buffer_reuse` (wrapper.py)
3. Add a new field `V.graph.inplaced_to_remove`.  For inplace buffers subject to remove, we don't actually remove them but put them in this dedicated set. This simplifies the life cycle management of inplace buffers. This set is used to 1) avoid unnecessary store in DeferredLine; 2) avoid alias var definitions in kernel (graph.py/scheduler.py).
4. Do not allocate buffer for `ExternKernelAlloc` and `MultiOutput` (wrapper.py) since these are pre-allocated.
5. Make sure reuse key matches on inplace buffer reuse in `allocate` (scheduler.py)
6. Do not inplace reuse buffers with layout `(ir.MultiOutputLayout, ir.MutationLayout, ir.AliasedLayout)`. This is a conservative choice and can revisit later if we can still inplace reuse buffers with some of these layouts.
7. Add kernel output buffers to `available_buffer_names` in `codegen` (scheduler.py) to make sure `remaining_uses` is correct in `allocate` (scheduler.py)

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
The approach seems right, but the test failures look real.

**Comment by jgong5** [comment]
@jansel There are two CI failures left. The unit test one seems not relevant to this change. For inductor_timm_training_2 gmixer_24_224 accuracy issue, weird that I cannot repro the failure locally, will check further, but feel free to let me know if you have any insights/hints. Thanks!

**Comment by jgong5** [comment]
> It the minifier works smoothly, you should be able to get a smaller reproducible graph which makes debugging much easier.

Thanks a lot for the tip, will have a try! BTW, seems trunk is broken? I got something like argument number not matching with all CUDA related test...

**Comment by soumith** [comment]
@jgong5 I believe that is because the `main` now requires a newer triton: https://github.com/pytorch/torchdynamo/pull/1585
Please upgrade your triton to the outlined commit and you should be good

**Comment by jgong5** [comment]
> It the minifier works smoothly, you should be able to get a smaller reproducible graph which makes debugging much easier

@desertfire Can this troubleshoot accuracy issues or only handle compilation/crash errors?

**Review by jansel — CHANGES_REQUESTED** [review]
Requesting changes to take this off my review queue, re-request review when tests pass.

**Comment by jgong5** [comment]
Thanks @desertfire @soumith for the hints on the minifier tool. It helped to generate a repro, even though not a small one since the issue only happened on the first clean-cache run. I managed to narrow it down manually to a smaller repro https://github.com/pytorch/torchdynamo/issues/1670. It seems an issue inside Triton compiler or the way how TorchInductor integrates Triton? @jansel

**Comment by jgong5** [comment]
@jansel Are you ok to land this PR with `inplace_buffers` set to `False` first? If it is turned on with the same PR change, only gmixer_24_224 training fails the accuracy test with the reason I explained in the description. I don't think the failure is caused by the change though but I will continue to check the root cause. Having it landed first allows others to experiment `inplace_buffers` if needed. If you are ok, the PR is in a good shape for review.

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
Yes, sounds good to me.
